src/main/java/org/jenkinsci/plugins/workflow/actions/ErrorAction.java

```diff
@@ -148,8 +149,28 @@ private static boolean equals(Throwable t1, Throwable t2) {
             return true;
         } else if (t1.getClass() != t2.getClass()) {
             return false;
+        } else if (!Objects.equals(t1.getMessage(), t2.getMessage())) {
+            return false;
         } else {
-            return Functions.printThrowable(t1).equals(Functions.printThrowable(t2));
+            // TODO: Steps which throw exceptions synchronously on the CPS VM thread break this equality check after
+            // a Jenkins restart. In that case, the original FlowNode gets an ErrorAction with the Throwable, which is
+            // written to disk, and then afterwards, when the error is thrown into the CPS-transformed script to resume
+            // execution, the location in the script and Continuable.SEPARATOR_STACK_ELEMENT are added to the stack
+            // trace of the same Throwable object using setStackTrace, so all subsequent serialized FlowNodes include
+            // those stack trace elements. This breaks stack trace equality, resulting in findOrigin returning the
+            // BlockEndNode of the parent of the step that threw the synchronous error.
+            // The `t1 == t2` reference equality check above masks the issue if you do not restart Jenkins.
```

Comment:
In particular, this affects the `error` step. I would like to fix this, but I couldn't really think of a straightforward way to do so, other than walking over `getStackTrace` with indices and ignoring missing `Continuable.SEPARATOR_STACK_ELEMENT` and `WorkflowScript.run` stack trace elements.

Reply:
The only thing I can think of is to inject an additional suppressed exception somewhere with a UUID, and then if two throwables both have that UUID somewhere in their cause/suppressed tree, they must be from a common origin.

Reply:
Hmm, yeah I guess when we create `ErrorAction` (in various places in `workflow-cps`), we could check if there is already a marker exception somewhere in its cause/suppressed tree, and if not, add one. Seems messy though. I will think about it.

Reply:
Tip: you can override `fillInStackTrace` to do nothing. Then the suppressed exception would add only one line to `Functions.printThrowable`. (If that is even used; not e.g. for `FlowInterruptedException` or `AbortExeption`.)